File: python/food-chain/food_chain.py
animals = ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"]
phrases = [
    "It wriggled and jiggled and tickled inside her.",
    "How absurd to swallow a bird!",
    "Imagine that, to swallow a cat!",
    "What a hog, to swallow a dog!",
    "Just opened her throat and swallowed a goat!",
    "I don't know how she swallowed a cow!",
    "She's dead, of course!",
]


def recite(start_verse, end_verse):
    ret = []
    for n in range(start_verse, end_verse + 1):
        ret.extend(verse(n) + [""])

    return ret[:-1]


def verse(n):
    ret = ["I know an old lady who swallowed a {}.".format(animals[n - 1])]
    if n > 1:
        ret.append("{}".format(phrases[n - 2]))
    if n < 8:
        for i in range(n - 1):
            ret.append("She swallowed the {} to catch the {}.".format(
                animals[n - i - 1], animals[n - i - 2]))
            if ret[-1].endswith("spider."):
                ret[-1] = ret[-1][:-1] + " that" + phrases[0][2:]
        ret.append("I don't know why she swallowed the fly. Perhaps she'll die.")

    return ret

File: haskell/food-chain/src/FoodChain.hs
module FoodChain
  ( song
  ) where

import           Data.List                      ( intercalate )

song :: String
song = intercalate "\n" (map verse [0 .. 7])

verse :: Int -> String
verse n =
  "I know an old lady who swallowed a "
    ++ animals
    !! n
    ++ ".\n"
    ++ verse'
    ++ verse''
 where
  verse' | n > 0     = phrases !! (n - 1)
         | otherwise = ""
  verse''
    | n < 7
    = concat (map f [0 .. n - 1])
      ++ "I don't know why she swallowed the fly. Perhaps she'll die.\n"
    | otherwise
    = ""
  f x
    | n - x == 2
    = "She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n"
    | otherwise
    = "She swallowed the "
      ++ animals
      !! (n - x)
      ++ " to catch the "
      ++ animals
      !! (n - x - 1)
      ++ ".\n"

animals :: [String]
animals = ["fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse"]

phrases :: [String]
phrases =
  [ "It wriggled and jiggled and tickled inside her.\n"
  , "How absurd to swallow a bird!\n"
  , "Imagine that, to swallow a cat!\n"
  , "What a hog, to swallow a dog!\n"
  , "Just opened her throat and swallowed a goat!\n"
  , "I don't know how she swallowed a cow!\n"
  , "She's dead, of course!\n"
  ]
